rust/lance-encoding/src/encodings/physical/bitpack.rs

```diff
@@ -0,0 +1,531 @@
+        Ok(EncodedBuffer {
+            bits_per_value: num_bits,
+            parts: packed_arrays,
+        })
+    }
+}
+
+fn pack_buffers(data: ArrayData, num_bits: u64, byte_len: usize) -> Vec<u8> {
+    let buffers = data.buffers();
+    let mut packed_buffers = vec![];
+    for buffer in buffers {
+        for i in 0..800 {
+            if i % 4 == 0 {
+                println!("{} = {}", i / 4, &buffer[i]);
+            }
+        }
+        let packed_buffer = pack_bits(&buffer, num_bits, byte_len);
+        packed_buffers.push(packed_buffer);
+    }
+    packed_buffers.concat()
+}
+
+fn pack_bits(src: &[u8], num_bits: u64, byte_len: usize) -> Vec<u8> {
+    // calculate the total number of bytes we need to allocate for the destination.
+    // this will be the number of items in the source array times the number of bits.
+    let src_items = src.len() / byte_len as usize;
+    let mut dst_bytes_total = src_items * num_bits as usize / 8;
+
+    // if if there's a partial byte at the end, we need to allocate one more byte
+    if (src_items * num_bits as usize) % 8 != 0 {
+        dst_bytes_total += 1;
+    }
+
+    let mut dst = vec![0u8; dst_bytes_total];
+    let mut dst_idx = 0;
+    let mut dst_offset = 0;
+    let bit_len = byte_len as u64 * 8;
+
+    let mut mask = 0u64;
+    for _ in 0..num_bits {
+        mask = mask << 1 | 1;
+    }
+
+    let mut src_idx = 0;
+    while src_idx < src.len() {
+        let mut curr_mask = mask;
+        let mut curr_src = src[src_idx] & curr_mask as u8;
+        let mut src_offset = 0;
+        let mut src_bits_written = 0;
+
+        while src_bits_written < num_bits {
+            dst[dst_idx] += (curr_src >> src_offset) << dst_offset;
+            let bits_written = (num_bits - src_bits_written)
+                .min(8 - src_offset)
+                .min(8 - dst_offset);
+            src_bits_written += bits_written;
+            dst_offset += bits_written;
+            src_offset += bits_written;
+
+            if dst_offset == 8 {
+                dst_idx += 1;
+                dst_offset = 0;
+            }
+
+            if src_offset == 8 {
+                src_idx += 1;
+                src_offset = 0;
+                curr_mask >>= 8;
+                if src_idx == src.len() {
+                    break;
+                }
+                curr_src = src[src_idx] & curr_mask as u8;
+            }
+        }
+
+        // advance source_offset to the next byte if we're not at the end..
+        // note that we don't need to do this if we wrote the full number of bits
+        // because source index would have been advanced by the inner loop above
+        if bit_len != num_bits {
+            let mut partial_bytes_written = num_bits / 8;
+
+            // if we didn't write the full byte for the last byte, increment by one because
+            // we wrote a partial byte
+            if bit_len % num_bits != 0 {
+                partial_bytes_written += 1;
+            }
```

Comment:
perhaps we can have a general `divide_round_up` function

Reply:
https://github.com/lancedb/lance/pull/2333#discussion_r1686656140